File: 2016/src/day21/task1/Task1.java
package day21.task1;

import java.nio.file.Files;
import java.nio.file.Path;
import java.time.Duration;
import java.time.Instant;

public class Task1 {
   Path path = Path.of("inputs", "day21", "input.txt");
   String start = "abcdefgh";

   void task() {
      try {
         var lines = Files.readAllLines(this.path);

         var sb = new StringBuilder();
         var result = this.start;
         for (var line : lines) {
            var segments = line.split(" ");
            if (line.startsWith("swap position")) {
               var x = Integer.parseInt(segments[2]);
               var y = Integer.parseInt(segments[5]);
               var idx1 = Integer.min(x, y);
               var idx2 = Integer.max(x, y);
               result = sb
                     .append(result.substring(0, idx1))
                     .append(result.charAt(idx2))
                     .append(result.substring(idx1 + 1, idx2))
                     .append(result.charAt(idx1))
                     .append(result.substring(idx2 + 1)).toString();
               sb.delete(0, sb.length());
               continue;
            }
            if (line.startsWith("swap letter")) {
               result = result.replace(segments[2], "#").replace(segments[5], segments[2]).replace("#", segments[5]);
               continue;
            }
            if (line.startsWith("reverse")) {
               var idx1 = Integer.parseInt(segments[2]);
               var idx2 = Integer.parseInt(segments[4]);
               var reversed = sb.append(result.substring(idx1, idx2 + 1)).reverse().toString();
               result = sb
                     .delete(0, sb.length())
                     .append(result.substring(0, idx1))
                     .append(reversed).append(result.substring(idx2 + 1))
                     .toString();
               sb.delete(0, sb.length());
               continue;
            }
            if (line.startsWith("rotate based")) {
               var idx = result.indexOf(segments[6]);

               var offset = result.length() - ((1 + idx + (idx >= 4 ? 1 : 0)) % result.length());
               result = sb
                     .append(result.substring(offset))
                     .append(result.substring(0, offset))
                     .toString();
               sb.delete(0, sb.length());
               continue;
            }
            if (line.startsWith("rotate")) {
               var offset = Integer.parseInt(segments[2]) % result.length();
               if (segments[1].equals("right")) {
                  offset = result.length() - offset;
               }

               result = sb
                     .append(result.substring(offset))
                     .append(result.substring(0, offset))
                     .toString();
               sb.delete(0, sb.length());
               continue;
            }
            if (line.startsWith("move")) {
               var from = Integer.parseInt(segments[2]);
               var to = Integer.parseInt(segments[5]);

               result = sb
                     .append(result.substring(0, from))
                     .append(result.substring(from + 1))
                     .insert(to, result.charAt(from))
                     .toString();
               sb.delete(0, sb.length());
               continue;
            }
            throw new RuntimeException("Unknown line: " + line);
         }
         System.out.println(result);
      } catch (Exception e) {
         e.printStackTrace();
         System.exit(1);
      }
   }

   public static void main(String[] args) {
      var start = Instant.now();

      new Task1().task();

      System.out.println("Done in " + Duration.between(start, Instant.now()).toString().substring(2).toLowerCase());
   }
}

File: 2017/src/day16/task1.ts
import path from "node:path";

const PATH = path.join("inputs", "day16", "input.txt");

async function task1() {
   let result = (await Bun.file(PATH).text())
      .trim()
      .split(",")
      .map((move) => [move.charAt(0), move.substring(1)] as [string, string])
      .reduce((acc, move) => {
         switch (move[0]) {
            case "s":
               const offset = +move[1];
               return acc.substring(acc.length - offset) + acc.substring(0, acc.length - offset);
            case "x":
               const positions = move[1]
                  .split("/")
                  .map(Number)
                  .sort((a, b) => a - b) as [number, number];
               return (
                  acc.substring(0, positions[0]) +
                  acc.charAt(positions[1]) +
                  acc.substring(positions[0] + 1, positions[1]) +
                  acc.charAt(positions[0]) +
                  acc.substring(positions[1] + 1)
               );
            case "p":
               const partners = move[1].split("/") as [string, string];
               return acc
                  .replace(partners[0], "#")
                  .replace(partners[1], partners[0])
                  .replace("#", partners[1]);
            default:
               throw new Error(`Unknown move: ${move[0]}`);
         }
      }, "abcdefghijklmnop");

   console.log(result);
}

console.time("Task Done");
await task1();
console.timeEnd("Task Done");
